### PR TITLE
Kube quadlets can support autoupdate as well as containers

### PR DESCRIPTION
**Pull request by rhatdan**
#### Does this PR introduce a user-facing change?

```release-note
Kubernetes quadlets now supports auto update.
```

### CODE REVIEW

File: docs/source/markdown/podman-systemd.unit.5.md [hunk 1]
```diff
@@ -121,7 +121,7 @@ Valid options for `[Container]` are listed below:
 | Rootfs=/var/lib/rootfs         | --rootfs /var/lib/rootfs                             |
 | Notify=true                    | --sdnotify container                                 |
 | PodmanArgs=--add-host foobar   | --add-host foobar                                    |
-| PublishPort=true               | --publish                                            |
+| PublishPort=50-59              | --publish 50-59                                      |
```
Comment (vrothberg) [inline docs/source/markdown/podman-systemd.unit.5.md:124]: This change seems unrelated. Can you break it out into a separate commit?

File: pkg/systemd/quadlet/quadlet.go [hunk 2]
```diff
@@ -961,6 +963,22 @@ func ConvertKube(kube *parser.UnitFile, isUser bool) (*parser.UnitFile, error) {
 
 	addNetworks(kube, KubeGroup, service, execStart)
 
+	updateMaps := kube.LookupAllStrv(KubeGroup, KeyAutoUpdate)
+	for _, update := range updateMaps {
+		val := strings.SplitN(update, "/", 2)
+		switch {
```
Comment (vrothberg) [inline pkg/systemd/quadlet/quadlet.go:969]: I'd prefer to just pass the values over without trying to validate them (i.e, compare to "registry" etc.).  Only `auto update ` should do that IMO to avoid scattering.

**Review by vrothberg — COMMENTED** [review]
Really nice idea!

@ygalblum PTAL

**Comment by TomSweeneyRedHat** [comment]
LGTM
but definitely want a @ygalblum head nod before merge

File: docs/source/markdown/podman-systemd.unit.5.md [hunk 3]
```diff
@@ -517,17 +517,28 @@ There is only one required key, `Yaml`, which defines the path to the Kubernetes
...
+| LogDriver=journald                  | --log-driver journald                            |
+| Network=host                        | --net host                                       |
+| PodmanArgs=\-\-annotation=key=value | --annotation=key=value                           |
+| PublishPort=59-60                   | --publish=59-60                                  |
+| UserNS=keep-id:uid=200,gid=210      | --userns keep-id:uid=200,gid=210                 |
+| Yaml=/tmp/kube.yaml                 | podman kube play /tmp/kube.yaml                  |
 
 Supported keys in the `[Kube]` section are:
 
+### `AutoUpdate=`
```
Comment (ygalblum) [inline docs/source/markdown/podman-systemd.unit.5.md:532]: Please add `This key may be used multiple times` to indicate that this key may be set multiple times to support multiple containers

**Comment by ygalblum** [comment]
LGTM, @vrothberg PTAL (following your comments)

File: docs/source/markdown/podman-systemd.unit.5.md [hunk 4]
```diff
@@ -517,17 +517,28 @@ There is only one required key, `Yaml`, which defines the path to the Kubernetes
...
+
+Indicates whether containers will be auto-updated ([podman-auto-update(1)](podman-auto-update.1.md)). AutoUpdate can be specified multiple times. The following values are supported:
+
+* `registry`: Requires a fully-qualified image reference (e.g., quay.io/podman/stable:latest) to be used to create the container. This enforcement is necessary to know which images to actually check and pull. If an image ID was used, Podman does not know which image to check/pull anymore.
+
+* `local`: Tells Podman to compare the image a container is using to the image with its raw name in local storage. If an image is updated locally, Podman simply restarts the systemd unit executing the Kubernetes quadlet.
+
+* `name/(local|registry)`: Tells Podman to preform the `local` or `registry` autoupdate on the specified container name.
```
Comment (vrothberg) [inline docs/source/markdown/podman-systemd.unit.5.md:540]:
```suggestion
* `name/(local|registry)`: Tells Podman to perform the `local` or `registry` autoupdate on the specified container name.
```

File: pkg/systemd/quadlet/quadlet.go (hunk 2 above)
Comment (vrothberg) [reply pkg/systemd/quadlet/quadlet.go]: @rhatdan @ygalblum this comment/issue is still open.  While I don't anticipate any new auto-update policies, I do think that scattered logic is imposing a risk when adding a new one.  In other words, IMO we should keep input _validation_ to a minimum in Quadlet.

File: pkg/systemd/quadlet/quadlet.go [hunk 5]
```diff
@@ -567,7 +569,7 @@ func ConvertContainer(container *parser.UnitFile, isUser bool) (*parser.UnitFile
 	update, ok := container.Lookup(ContainerGroup, KeyAutoUpdate)
 	if ok && len(update) > 0 {
 		podman.addLabels(map[string]string{
-			"io.containers.autoupdate": update,
+			define.AutoUpdateLabel: update,
```
Comment (Luap99) [inline pkg/systemd/quadlet/quadlet.go:572]: importing define seems to bloat this a lot, I agree that we should not duplicate strings so you have to move the definition in another package

**Review by Luap99 — COMMENTED** [review]
Why is this bloat approved, looks like this bloats quadlet by 800KB? I think the goal is to keep quadlet as small as possible so you should fix this.

File: pkg/systemd/quadlet/quadlet.go (hunk 5 above)
Comment (rhatdan) [reply pkg/systemd/quadlet/quadlet.go:572]: I think we should examine why define is bloating.  Quadlet size does not make much difference.

File: pkg/systemd/quadlet/quadlet.go [hunk 6]
```diff
@@ -961,6 +963,23 @@ func ConvertKube(kube *parser.UnitFile, isUser bool) (*parser.UnitFile, error) {
...
+		val := strings.SplitN(update, "/", 2)
+		if len(val) == 2 {
+			annotation = annotation + "/" + val[0]
+			updateType = val[1]
+		}
+		switch updateType {
+		case "local", "registry":
+			execStart.addf("%s=%s", annotation, updateType)
+		default:
+			return nil, fmt.Errorf("invalid %s value: %s", KeyAutoUpdate, update)
```
Comment (vrothberg) [inline pkg/systemd/quadlet/quadlet.go:979]: My previous comment is still open (see https://github.com/containers/podman/pull/18975#pullrequestreview-1493469714).  I don't think we should validate input here.  Code scattering comes at a risk.

**Review by vrothberg — COMMENTED** [review]
One comment remains open.  Could you could squash the commits into one?

File: pkg/systemd/quadlet/quadlet.go (hunk 5 above)
Comment (Luap99) [reply pkg/systemd/quadlet/quadlet.go:572]: The size will matter very much, literally the first sentence in in the systemd man page:
> Generators are small executables ...

The generator blocks early boot, the more bloat we have the slower the binary will run and thus result in slower boot times.
Based on a quick benchmark which just uses -dryrun it doubles the execution time.

main:
```
$ hyperfine --warmup 10  -N 'bin/quadlet-main -dryrun'
Benchmark 1: bin/quadlet-main -dryrun
  Time (mean ± σ):       1.4 ms ±   0.3 ms    [User: 0.4 ms, System: 1.2 ms]
  Range (min … max):     1.1 ms …   4.3 ms    1969 runs
```
this PR:
```
$ hyperfine --warmup 10  -N 'bin/quadlet -dryrun'
Benchmark 1: bin/quadlet -dryrun
  Time (mean ± σ):       3.0 ms ±   0.5 ms    [User: 1.2 ms, System: 2.0 ms]
  Range (min … max):     2.5 ms …   7.2 ms    885 runs
```

So yeah sure 1.5 ms is not much but the whole point of the bloat check it to catch these things and not ignore them.

**Comment by Luap99** [comment]
Also can you please do not commit these bogus github commits. https://github.com/containers/podman/pull/18975/commits/c59ea124b6c6cb79ec79ceff6c1fa4ab4d89c8a6
They contain zero context and fill the git history with pointless commits so please squash them locally.

File: pkg/systemd/quadlet/quadlet.go (hunk 5 above)
Comment (rhatdan) [reply pkg/systemd/quadlet/quadlet.go:572]: I agree but we need to figure how how libpod/define is causing bloat.  The define section is supposed to be used to prevent bloat.  I can fix it hear and sweep it under the rug, but until someone looks at define to see where it got bloat...

**Comment by baude** [comment]
LGTM
